File: apps/java/src/main/java/com/functions/ExternalDependencyFunction.java
package com.functions;

import com.microsoft.azure.functions.*;
import com.microsoft.azure.functions.annotation.*;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Optional;

/**
 * External dependency probe — measures latency to an external URL.
 */
public class ExternalDependencyFunction {

    private static final HttpClient HTTP_CLIENT = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();

    @FunctionName("externalDependency")
    public HttpResponseMessage run(
            @HttpTrigger(
                name = "req",
                methods = {HttpMethod.GET},
                authLevel = AuthorizationLevel.ANONYMOUS,
                route = "dependency")
            HttpRequestMessage<Optional<String>> request,
            final ExecutionContext context) {

        String url = request.getQueryParameters().getOrDefault("url", "https://httpbin.org/get");
        context.getLogger().info("External dependency check: " + url);

        long start = System.currentTimeMillis();
        try {
            HttpRequest httpReq = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();

            HttpResponse<String> response = HTTP_CLIENT.send(httpReq, HttpResponse.BodyHandlers.ofString());
            long elapsed = System.currentTimeMillis() - start;

            String body = String.format(
                "{\"url\":\"%s\",\"statusCode\":%d,\"latencyMs\":%d,\"success\":true}",
                url, response.statusCode(), elapsed);

            return request.createResponseBuilder(HttpStatus.OK)
                    .header("Content-Type", "application/json")
                    .body(body)
                    .build();
        } catch (Exception e) {
            long elapsed = System.currentTimeMillis() - start;
            String body = String.format(
                "{\"url\":\"%s\",\"error\":\"%s\",\"latencyMs\":%d,\"success\":false}",
                url, e.getMessage(), elapsed);

            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                    .header("Content-Type", "application/json")
                    .body(body)
                    .build();
        }
    }
}

File: apps/dotnet/Functions/ExternalDependencyFunction.cs
using System.Diagnostics;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;

namespace AzureFunctionsGuide.Functions;

/// <summary>
/// External HTTP dependency call — measures latency to httpbin.org.
/// Route: GET /api/dependency
/// </summary>
public class ExternalDependencyFunction
{
    private readonly ILogger<ExternalDependencyFunction> _logger;
    private static readonly HttpClient _httpClient = new();
    private const string ExternalUrl = "https://httpbin.org/get";

    public ExternalDependencyFunction(ILogger<ExternalDependencyFunction> logger)
    {
        _logger = logger;
    }

    [Function("externalDependency")]
    public async Task<IActionResult> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "dependency")] HttpRequest req)
    {
        _logger.LogInformation("External dependency endpoint requested (url={Url})", ExternalUrl);

        var sw = Stopwatch.StartNew();
        try
        {
            var response = await _httpClient.GetAsync(ExternalUrl);
            sw.Stop();

            _logger.LogInformation("External call completed (status={Status}, time={Elapsed}ms)",
                (int)response.StatusCode, sw.ElapsedMilliseconds);

            return new OkObjectResult(new
            {
                url = ExternalUrl,
                statusCode = (int)response.StatusCode,
                latencyMs = sw.ElapsedMilliseconds,
                success = response.IsSuccessStatusCode
            });
        }
        catch (Exception ex)
        {
            sw.Stop();
            _logger.LogError(ex, "External call failed (time={Elapsed}ms)", sw.ElapsedMilliseconds);

            return new ObjectResult(new
            {
                url = ExternalUrl,
                statusCode = 0,
                latencyMs = sw.ElapsedMilliseconds,
                success = false,
                error = ex.Message
            })
            { StatusCode = 502 };
        }
    }
}
